classes/sync.php

```diff
@@ -244,22 +250,32 @@ public function sync(progress_trace $trace) {
                             continue;
                         }
                         $foundexternalcohort = true;
-                        if (!empty($currentusers[$fields[$remoteuserfieldl]])) {
+                        $remotefield = strtolower($fields[$remoteuserfieldl]);
+                        if (!empty($currentusers[$remotefield])) {
                             // This user is already a member of the cohort.
-                            unset($currentusers[$fields[$remoteuserfieldl]]);
+                            unset($currentusers[$remotefield]);
                         } else {
                             // Add user to cohort.
                             $newmember = new stdClass();
                             $newmember->cohortid  = $cohort->id;
                             $newmember->userid    = $fields[$remoteuserfieldl];
                             $newmember->timeadded = time();
                             $newmembers[] = $newmember;
-                            $needusers[] = $fields[$remoteuserfieldl];
+                            $needusers[] = $remotefield;
                         }
                     }
                 }
             }
             if ($foundexternalcohort && empty($removeaction) && !empty($currentusers)) {
+                if (!empty($this->config->maxremovals) && $countdeletes > $this->config->maxremovals) {
+                    $message = "The cohort sync process has removed $countdeletes members,
+                                this exceeds the max removal threshold of {$this->config->maxremovals} so the process was stopped.
+                                If this is unexpected you should check the validity of the data or it will continue removal on next cron.";
+                    $this->email_admins($message);
+                    mtrace($message);
+                    $trace->finished();
+                    return 1;
+                }
```

Comment:
This looks like if you exceeded the max removal threshold it would only be caught if you had multiple cohorts, this won't catch removals that happen within a single cohort

Reply:
yeah - that was intentional - I'll update the strings a bit to see if I can make that clear - I was thinking that it's better to "finish" dealing with one cohort - rather than leave it in a partially synced state.

Reply:
I ended up adding the count of the items about to be deleted in this cohort which should also help. This does have the after affect that if it's set too low, then it will never get cleared until someone increases the maxremoval threshold but I think that's ok.